lib/mayaUsd/nodes/layerManager.cpp

```diff
@@ -414,40 +450,142 @@ bool LayerDatabase::saveUsd(bool isExport)
 
         if (result == kAbort) {
             return false;
-        } else if (result == kCompleted) {
+        } else if (result == kCompleted && _internalProxiesToSave.length() == 0) {
+            return true;
+        } else if (result == kPartiallyCompleted && !getProxiesToSave(isExport)) {
             return true;
-        } else if (result == kPartiallyCompleted) {
-            if (!getStagesToSave(isExport)) {
-                return true;
-            }
         }
 
-        if (MayaUsd::kAbort != result) {
-            if (MayaUsd::utils::kSaveToUSDFiles == opt) {
-                result = saveUsdToUsdFiles();
-            } else {
-                result = saveUsdToMayaFile();
-            }
+        if (MayaUsd::utils::kSaveToUSDFiles == opt) {
+            result = saveUsdToUsdFiles();
+        } else {
+            result = saveUsdToMayaFile();
         }
     } else {
         result = MayaUsd::kCompleted;
     }
 
     _proxiesToSave.clear();
+    _internalProxiesToSave.clear();
     return (MayaUsd::kCompleted == result);
 }
 
-BatchSaveResult LayerDatabase::saveUsdToMayaFile()
+MStatus addLayerToBuilder(
+    MayaUsd::LayerManager* lm,
+    MArrayDataBuilder&     builder,
+    SdfLayerHandle         layer,
+    bool                   isAnon,
+    bool                   stubOnly = false,
+    bool                   exportOnlyIfDitry = false)
 {
-    MayaUsd::LayerManager* lm = findNode();
+    if (!lm)
+        return MS::kFailure;
+
+    MStatus     status = MS::kSuccess;
+    MDataHandle layersElemHandle = builder.addLast(&status);
+    CHECK_MSTATUS_AND_RETURN_IT(status);
+    MDataHandle idHandle = layersElemHandle.child(lm->identifier);
+    MDataHandle serializedHandle = layersElemHandle.child(lm->serialized);
+    MDataHandle anonHandle = layersElemHandle.child(lm->anonymous);
+
+    idHandle.setString(UsdMayaUtil::convert(layer->GetIdentifier()));
+    anonHandle.setBool(isAnon);
+
+    std::string temp;
+    if (!stubOnly && ((exportOnlyIfDitry && layer->IsDirty()) || !exportOnlyIfDitry)) {
+        if (!layer->ExportToString(&temp)) {
+            status = MS::kFailure;
+        }
+    }
+
+    serializedHandle.setString(UsdMayaUtil::convert(temp));
```

Comment:
Should we be setting this when status == MS:kFailure (from line 497). Won't temp be bad?

Reply:
Yes, but this is the old behavior, I didn't want to change it.. But I agree with you

Reply:
Let me know if I should change it